src/vsr/replica.zig

```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
 
-            self.op = latest.op;
+            const previous_op = self.op;
+            const previous_commit_max = self.commit_max;
+
+            self.op = op;
+            self.journal.remove_entries_from(self.op + 1);
+
             // Crucially, we must never rewind `commit_max` (and then `commit_min`) because
             // `commit_min` represents what we have already applied to our state machine:
-            self.commit_max = std.math.max(self.commit_max, k);
+            self.commit_max = std.math.max(self.commit_max, commit_max);
 
             assert(self.commit_min <= self.commit_max);
-            assert(self.op >= self.commit_max);
+            assert(self.commit_max <= self.op);
 
-            // Do not set the latest op as dirty if we already have it exactly:
-            // Otherwise, this would trigger a repair and delay the view change, or worse, it would
-            // prevent us from assisting another replica to recover when we do in fact have the op.
-            if (self.journal.has(latest)) {
-                log.debug("{}: {s}: latest op exists exactly", .{ self.replica, method });
-            } else {
-                self.journal.set_header_as_dirty(latest);
+            log.debug("{}: {s}: view={} op={}..{} commit={}..{}", .{
+                self.replica,
+                method,
+                self.view,
+                previous_op,
+                self.op,
+                previous_commit_max,
+                self.commit_max,
+            });
+        }
+
+        fn set_log_from_do_view_change_messages(self: *Self) void {
+            assert(self.status == .view_change);
+            assert(self.leader_index(self.view) == self.replica);
+            assert(self.replica_count > 1);
+            assert(self.start_view_change_quorum);
+            assert(self.do_view_change_quorum);
+
+            const do_view_change_head = self.do_view_change_quorum_head();
+            assert(do_view_change_head.view_normal >= self.view_normal);
+            assert(do_view_change_head.op >= self.commit_min);
+            assert(do_view_change_head.op >= do_view_change_head.commit_min_max);
+            assert(do_view_change_head.commit_min_max >= self.commit_min);
+
+            const view_normal_canonical = do_view_change_head.view_normal;
+            // `op_uncanonical` must be computed before calling `set_op_and_commit_max()`, since
+            // that may change `replica.op`.
+            //
+            // Don't remove the uncanonical headers yet, though — even though the removed headers are
+            // a subset of the DVC headers, removing and then adding them back would cause clean
+            // headers to become dirty.
+            const op_uncanonical = self.op_uncanonical_min(view_normal_canonical);
+            assert(op_uncanonical <= self.op + 1);
+            assert(op_uncanonical >= self.op + 1 -| config.pipeline_max);
+            assert(op_uncanonical >= self.commit_min);
```

Comment:
Should it not always be > `self.commit_min`?

Reply:
i.e. Can `commit_min` itself be uncanonical?

Reply:
Good point; `commit_min` is always canonical. `op_uncanonical_min` already had the correct check, too.